### ISSUE TITLE
Investigate timing issues with moment creation

### ISSUE THREAD

**Issue by mj1856** [moment/moment#2659]
This isn't necessarily a *bug*, but a [StackOverflow user stumbled across an interesting effect](http://stackoverflow.com/a/33003021/634824).

Rapid creation of two `Date` objects can go through several thousand iterations before getting different values: 
```javascript
setInterval(function (){console.log(new Date().getTime() === new Date().getTime());}, 0);
```
![image](https://cloud.githubusercontent.com/assets/1396388/10354742/f8cd389a-6d17-11e5-81ae-ffe734550ffb.png)

But rapid creation of two `moment` objects gets different values after under 100 iterations:
```javascript
setInterval(function (){console.log(moment().isSame(moment()));}, 0);
```
![image](https://cloud.githubusercontent.com/assets/1396388/10354696/7dbceef2-6d17-11e5-8503-7b049a1f05ab.png)

Is this just because a `moment` is more expensive than a `Date` in terms of memory and CPU?  Is there perf work that could be done here?  Does it matter?

**Comment by 1337** [comment]
Hey mate, I wrote that SO answer, and it was not intended to address what you see from `moment.isSame()`.

You must know what `Date.getTime()` does: get the unix timestamp of the time the `Date` contains. In a line like `new Date().getTime() === new Date().getTime()` (*sic.*), the two values can be different if the first construction and the second construction crosses the millisecond boundary.

**Comment by mj1856** [comment]
Yep, I know that. It's the comments that were interesting.  Thanks.

I'm just wondering if we can make moment perform a bit better.  There may be a better way to measure than this.

**Comment by marwahaha** [comment]
Can this one be up-for-grabs?

**Comment by tomerle** [comment]
Hi @marwahaha, I think I can boost the comparison methods. Are you guys looking for help on this one?

**Comment by marwahaha** [comment]
@tomerle - go for it. Even if you provide testing/benchmarking methods, that would be useful

**Comment by tomerle** [comment]
The PR adds benchmarks for `get` and `set`. These are pretty similar to the existing benchmarks for `startOf` and `endOf`, in the sense that they compare the same method across the different units. I also added a benchmark to compare all of the compare methods we're discussing here to each other.  
  
Two conclusions I got from benching `get` and `set` were:
1. Handling `week` & `isoWeek` is slower.
2. Handling `date` is faster.
  
I started with the 2nd one because it seems odd to me that just delegating to the native `Date` methods is so much slower in this case. I tried out some stuff and seems like avoiding the string concatenations in `get-set.js` does the trick. Rewriting the file with switch-case statements proved to be the fastest implementation per my benching, bringing all the units to about the same speed of `date`. Since a lot of the other functionality relies on this file, the improvement is visible in other benchmarks too.  
  
For now, I'll only add a link to a branch on my fork with the above change to `get-set.js`, so everyone can see it. LMK if you prefer to merge this one first, or both together, or none, ... I have some more ideas I'm working on that seem worth considering afterwards.

**Comment by tomerle** [comment]
This is the link to the `get-set.js` suggestion: https://github.com/tomerle/moment/commit/83ffefabc04fa2b417185d6375f20e51e7c2de3e  
For me the compare methods got about twice as fast, and `get` had impressing results...
  
Before:
```
>> get second x 3,733,522 ops/sec ±0.98% (98 runs sampled)
>> get minute x 2,300,286 ops/sec ±0.23% (99 runs sampled)
>> get hour x 2,507,760 ops/sec ±0.59% (102 runs sampled)
>> get date x 2,684,643 ops/sec ±0.68% (101 runs sampled)
>> get day x 10,807,065 ops/sec ±0.20% (100 runs sampled)
>> get isoWeek x 137,027 ops/sec ±0.21% (99 runs sampled)
>> get week x 142,939 ops/sec ±0.31% (95 runs sampled)
>> get month x 2,682,240 ops/sec ±1.31% (92 runs sampled)
>> get quarter x 2,776,126 ops/sec ±0.99% (98 runs sampled)
>> get year x 2,801,299 ops/sec ±1.04% (98 runs sampled)
```

After:
```
>> get second x 36,254,565 ops/sec ±0.21% (100 runs sampled)
>> get minute x 10,065,896 ops/sec ±0.33% (102 runs sampled)
>> get hour x 10,327,325 ops/sec ±0.16% (100 runs sampled)
>> get date x 10,295,007 ops/sec ±0.39% (102 runs sampled)
>> get day x 10,505,391 ops/sec ±0.31% (102 runs sampled)
>> get isoWeek x 236,274 ops/sec ±0.80% (95 runs sampled)
>> get week x 265,767 ops/sec ±0.22% (98 runs sampled)
>> get month x 10,525,210 ops/sec ±1.39% (98 runs sampled)
>> get quarter x 9,836,370 ops/sec ±0.23% (97 runs sampled)
>> get year x 9,649,852 ops/sec ±0.14% (101 runs sampled)
```
  
The drop from 36 to 10 million ops/sec seems to be V8 deoptimization related, but I'm not sure we can gain something here, and for now I wanna concentrate on the compare methods. I wanna try using `get` instead of `startOf` & `endOf` and see how much we gain from it together with this.

**Comment by marwahaha** [comment]
Thanks for investigating @tomerle .
Is there anything we should change to Moment.js ?

**Comment by tomerle** [comment]
Hi @marwahaha,

I actually continued the work a bit back then and got to the point where all get methods were improved including the week related ones and even the compare methods. So yes... I'd love to rebase and send pull requests 👍 

Since I'm new to the project and not a master of time, I'll PR in small chunks and start with the get methods performance improvement. It's very simple and impacts other methods that rely on get too. I just need to rebase first and see if everything is still relevant. I'll try to have a PR opened and linked here this week...

**Comment by tomerle** [comment]
Hi again @marwahaha,
Tested my code again and created PR #5718
